### PR TITLE
Delete related opt-in tokens after a password change

### LINKED ISSUE #7510: Changing the password should invalidate password reset tokens
### Affected version(s)

all

### Description

Vulnerability Name: Broken Authentication – Session Token bug

I found a broken authentication vuln

POC:

1- Create an account
2- Confirm your email
3- Now request a password reset.
4- Don’t use the password reset link that was sent to your email.
5- Login to your account, remember don’t use the reset password link you requested in 3 step
6- Change your password in the Account Settings
7. After you change your password inside your account, Check the reset password link you requested in Step 3 in your email.
8. Change your password using the reset password token you requested.

Impact
token should expire
If the site has a token issue, The result is the reset password token in Step 3 is still usable and has not expired yet. Not invalidating the session token for the reset password is not a good practice for a company.

### PR DESCRIPTION
**Pull request by bytehead**
Fixes #7510

### CODE REVIEW

**Comment by bytehead** [comment]
Not sure, if this is the right way to fix this 🙃 
I thought about to extend the `OptInInterface` but that would be a BC break.

**Review by m-vo — DISMISSED** [review]
I agree with the reasoning for using the model. 

Ideally, there should be a service that has an `invalidate()` method (or something alike), that then operates on the underlying data structure, IMHO. Then, things would be properly encapsulated. But we can probably only do this for Contao 6 as this would mean rearranging a lot of things.

**Review by fritzmg — COMMENTED** [review]
We don't really need the adapters in the legacy classes not covered by tests. Directly using `OptInModel::deleteByTableAndEmail(…)` is fine there too.

**Review by m-vo — DISMISSED** [review]

**Review by fritzmg — DISMISSED** [review]

**Review by aschempp — CHANGES_REQUESTED** [review]
This should use the `relId` for a relation to the member and not the `email` that it was sent to. The email of the user could have been changed in the meantime.

**Comment by bytehead** [comment]
@aschempp changed in 1dd7446d89.

**Comment by bytehead** [comment]
Failing tests are unrelated.

File: core-bundle/src/Resources/contao/models/OptInModel.php [hunk 1]
```diff
@@ -228,4 +228,23 @@ public function setRelatedRecords(array $arrRelated)
...
+	 * @param int $intId
+	 *
+	 * @return integer The number of affected rows
+	 */
+	public static function deleteByTableAndId($strTable, $intId)
+	{
+		$t = static::$strTable;
+
+		return Database::getInstance()
+			->prepare("DELETE FROM $t WHERE $t.id = ? $t.confirmedOn = 0 AND $t.id IN (SELECT pid FROM tl_opt_in_related WHERE relTable=?)")
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/models/OptInModel.php]:
```suggestion
			->prepare("DELETE FROM $t WHERE $t.id = ? AND $t.confirmedOn = 0 AND $t.id IN (SELECT pid FROM tl_opt_in_related WHERE relTable=?)")
```

**Review by aschempp — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/models/OptInModel.php (hunk 1 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/models/OptInModel.php]: This still looks wrong though. I don't think we should query for ID but for relId in the subtable?

**Review by aschempp — DISMISSED** [review]

File: core-bundle/src/Resources/contao/models/OptInModel.php (hunk 1 above)
Comment (bytehead) [reply core-bundle/src/Resources/contao/models/OptInModel.php]: I do query the subtable: 

`$t.id IN (SELECT pid FROM tl_opt_in_related WHERE relTable=?)")`

Comment (aschempp) [reply core-bundle/src/Resources/contao/models/OptInModel.php]: yes but the member ID is in `tl_opt_in_related.relId` and not in `tl_opt_in.id`

Comment (bytehead) [reply core-bundle/src/Resources/contao/models/OptInModel.php]: Like in f5f4b680d2?

**Comment by leofeyer** [comment]
Shouldn‘t we rather have a `findUnconfirmedByRelatedTableAndId()` method?

```php
if ($models = OptInModel::findUnconfirmedByRelatedTableAndId('tl_member', $objMember->id)) {
    foreach ($models as $model) {
        $model->delete();
    }
}
```

We don‘t use `deleteBy…` methods anywhere else.

**Comment by bytehead** [comment]
Valid point 👍🏼 (it is an legacy model thing anyway) - I'll update it accordingly.

**Comment by bytehead** [comment]
Instead of a one-liner in each module it's 7 lines each now 🤷🏼‍♂️

**Comment by leofeyer** [comment]
Yes, but it's also more stringent that way IMHO.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bytehead.

**Pull request merged**